### DISCUSSION TITLE
Is there a way to mock data for all inserts and fetches implicitly using a mock database adapter?

### DISCUSSION THREAD

**Discussion by nelsnelson** [jeremyevans/sequel#2095] [General]
Greetings,-

I've mistakenly opened an issue in an attempt to resolve the issue I had been experiencing trying to configure a mock database adapter for use in a spec test suite which will support automatically incremented primary key id values for row fetches (and inserts).

It now seems clear that I had a fundamental misunderstanding of how the mock database adapter is intended to be operated within the context of an rspec test suite.

I may still not have the right picture here.  I've been hoping to use the mock adapter in a testing configuration (`ENV['APP_ENV'] = 'test'`) which would seamlessly support existing code which set up database connection for model classes.

Is there a way to configure the mock adapter to always generate an automatically incremented id for fetches and inserts?

One example of an error I'm trying to avoid:

```
Failure/Error: add_example_relation(ExampleRelation.find_or_create(name: 'example-relation-01'))

Sequel::Error:
       model object #<Example @values={:name=>"example-instance-01"}> does not have a primary key
```

My thoughts were that this was happening because the mock database adapter wasn't correctly assigning an `:id=>value` where `value` was an automatically incremented counter internal to the adapter.  But it is [beginning to dawn on me](https://github.com/jeremyevans/sequel/issues/2094#issuecomment-1806098396) that it is instead because I'm not explicitly invoking the `Database#autoid=` and `Database#fetch=` methods for the database following its initialization.

Is there an alternative to explicit invocation?  Perhaps it can be accomplished at a configuration level using a connection string parameter?

I'm guessing such functionality for returning generated mock data for fetches and inserts based on the schema of the individual model for a particular class inherited from `Sequel::Model` is not a feature of the mock adapter.  The challenges for such a feature would probably involve some prohibitive and perhaps even unnecessary complexity.

**Comment by jeremyevans** [comment]
The mock adapter is as the name says, a mock. The main purpose of it is testing Sequel itself.  It is not an in-memory database.  You may want to use an SQLite memory database if you want an in-memory database. Note that you can pass options when creating the mock Database (such as :fetch as :autoid).  If you are looking for examples of how to use the mock adapter, you probably want to look at Sequel's tests.

**Reply by nelsnelson** [reply]
I was beginning to consider the possibility that the mock was specific to Sequel testing.  You have now crystallized this for me.

It's now clear I've been trying to shoehorn the mock adapter into the wrong use case.
